inject.go

```diff
@@ -0,0 +1,235 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/ghodss/yaml"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	yamlDecoder "k8s.io/apimachinery/pkg/util/yaml"
+
+	// https://github.com/kubernetes/client-go/blob/master/INSTALL.md
+	"k8s.io/client-go/pkg/api/v1"
+	batch "k8s.io/client-go/pkg/apis/batch/v1"
+	"k8s.io/client-go/pkg/apis/extensions/v1beta1"
+)
+
+/*
+	Inject in init-container to set up transparent proxying
+	via iptables rules which send traffic to daemonset linkerd
+
+	This file based on:
+	https://github.com/istio/pilot/blob/master/platform/kube/inject/inject.go
+	and https://github.com/istio/pilot/tree/master/cmd/istioctl/inject.go
+
+  This code inserts the following into the pod specs:
+
+   spec:
+     initContainers:
+       - name: init
+         image: buoyantio/istio-init:v1
+         args:
+           - -p
+           - "4140"
+         imagePullPolicy: Always
+         securityContext:
+           capabilities:
+             add:
+             - NET_ADMIN
+*/
+
+const (
+	istioSidecarAnnotationSidecarKey   = "alpha.istio.io/linkerd-daemonset"
+	istioSidecarAnnotationSidecarValue = "injected"
+	initContainerName                  = "init-linkerd"
+	initImage                          = "buoyantio/istio-init:v1"
+)
+
+type Params struct {
+	LinkerdDaemonsetPort string
+}
+
+func dieIf(err error) {
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+}
+
+func injectIntoPodTemplateSpec(p *Params, t *v1.PodTemplateSpec) error {
+	// from https://github.com/istio/pilot/blob/master/platform/kube/inject/inject.go
+
+	if t.Annotations == nil {
+		t.Annotations = make(map[string]string)
+	} else if _, ok := t.Annotations[istioSidecarAnnotationSidecarKey]; ok {
+		// Return unmodified resource if init has already happened
+		return nil
+	}
+	t.Annotations[istioSidecarAnnotationSidecarKey] = istioSidecarAnnotationSidecarValue
+
+	// init-container
+	var annotations []interface{}
+	if initContainer, ok := t.Annotations["pod.beta.kubernetes.io/init-containers"]; ok {
+		if err := json.Unmarshal([]byte(initContainer), &annotations); err != nil {
+			return err
+		}
+	}
+	initArgs := []string{
+		"-p", p.LinkerdDaemonsetPort,
+	}
+	annotations = append(annotations, map[string]interface{}{
+		"name":            initContainerName,
+		"image":           initImage,
+		"args":            initArgs,
+		"imagePullPolicy": "Always",
+		"env": []map[string]interface{}{
+			map[string]interface{}{
+				"name": "NODE_NAME",
+				"valueFrom": map[string]interface{}{
+					"fieldRef": map[string]interface{}{
+						"apiVersion": "v1", // https://github.com/kubernetes/kubernetes/issues/39189
+						"fieldPath":  "spec.nodeName",
+					},
+				},
+			},
+		},
```

Comment:
Yeahhhh agreed. Since minikube has one node, maybe we could use the l5d service VIP in the iptables rules rather than the host ip (which we can't get in minikube)?

So I can get this to run in minikube by doing the following, though it's a bit much to ask a user to do this:

Modify prepare_proxy.sh to use DNAT to l5d service VIP:
```
iptables -t nat -A OUTPUT -p tcp -j DNAT --to ${L5D_SERVICE_HOST}:${LINKERD_PORT}                  -m comment --comment "istio/dnat-to-daemonset-l5d"
```

Build a linkerd/istio-init docker image locally and use that.
Use inject.go or manually insert the initContainer section in your pod spec. Make sure to remove` imagePullPolicy: Always` from this section though.

Deploy your modified yaml. Then:
```
L5D_PORT=$(kubectl get svc l5d -o jsonpath="{.spec.ports[0].nodePort}")
curl -v $(minikube ip):$L5D_PORT -H "Host: hello"
```

I was just talking to @adleong about this and maybe we can add a flag to `inject.go` so that people can run in minikube without having to build their own image for this.